ultravox/model/hubert.py

```diff
@@ -0,0 +1,119 @@
+from pathlib import Path
+
+import fairseq
+import joblib
+import torch
+from einops import rearrange
+from einops import repeat
+from torch import nn
+from torchaudio.functional import resample
+
+_DEFAULT_OUTPUT_LAYER = 9
+_DEFAULT_TARGET_SAMPLE_HZ = 16000
+# From the paper, 20 ms at 16 kHz.
+_DOWNSAMPLE_FACTOR = 320
+
+
+def _round_down_nearest_multiple(num, divisor):
+    return num // divisor * divisor
+
+
+def _curtail_to_multiple(t, mult, from_left=False):
+    data_len = t.shape[-1]
+    rounded_seq_len = _round_down_nearest_multiple(data_len, mult)
+    seq_slice = (
+        slice(None, rounded_seq_len) if not from_left else slice(-rounded_seq_len, None)
+    )
+    return t[..., seq_slice]
+
+
+def _exists(val):
+    return val is not None
+
+
+class HubertWithKmeans(nn.Module):
+    """
+    Adapted from @lucidrains
```

Comment:
Q: Does this mean Copyright notice should be copied over?
Mainly asking because the copyright stuff is very foreign to me, so if you know we don't need to I'd love to learn!

Reply:
IANAL, but that repo is also MIT licensed, so I don't think any additional notices need to be copied over.

Reply:
Yes, it's MIT so it should be fine to use. I think the way licenses/credits are handled varies by project/repo. Sometimes rigor around putting third party derived code in a separate dir with a LICENSE file, etc. I think given the early stage of this project, one lightweight idea could be to start a `CREDITS.md` file or similar and as new items come up, add a bullet on directories/files, link to source, and state the license.